[WIP][Proposal] PARQUET-2430: Add parquet joiner

MaxNevermind opened the pull request:
**_This PR is a proposal and Work In Progress._**


## Overview
ParquetJoiner feature is similar to ParquetRewrite class. ParquetRewrite allows to stitch files with the same schema into a single file while ParquetJoiner should enable stitching files with different schemas into a single file. That is possible when: 1) the number of rows in the main and extra files is the same, 2) the ordering of rows in the main and extra files is the same. Main benefit of ParquetJoiner is performance, for the cases when you join/stitch Terabytes/Petabytes of data that seemingly simple low level API can be up to 10x more resource efficient. 

## Implementation details
ParquetJoiner allows to specify the main input parquet file and extra input parquet files. ParquetJoiner will copy the main input as binary data and write extra input files with row groups adjusted to the main input. If main input is much larger than extra inputs then a lot of resources will be saved by working with the main input as binary.

## Use-case examples
A very large Parquet based dataset(dozens or hundreds of fields/Terabytes of data daily/Petabytes of historical partitions). The task is to modify a column or add a new column to it for all the historic data. It is trivial using Spark, but taking into consideration the share scale of a dataset it will take a lot of resources to do that. 

### Side notes
Note that this class of problems could be in theory solved by storing main input and extra inputs in HMS/Iceberg bucketed tables and use a view that joins those tables on the fly into the final version but in practice there is often a requirement to merge parquet files and have a single parquet sources in the file system.

## Use-case implementation details using Apache Spark
You can use Apache Spark to perform the join with ParquetJoiner, read the large main input and prepare the right side of a join in a way that each file on the left have a corresponding file on the right and it preserves records ordering on the right side in the same order as on the left side, that allows the whole input on the left and right to have the same number of files and the same number of records in corresponding files and the same ordering of records in each file pair. Then run ParquetJoiner in parallel for each file pair and perform a join. Example of the code that utilizes this new feature: https://gist.github.com/MaxNevermind/0feaaf380520ca34c2637027ef349a7d.

 


Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [PARQUET-2430: Add ParquetJoiner feature](https://issues.apache.org/jira/browse/PARQUET-2430) 

### Tests

- [x] My PR adds the following unit tests:
  -   testStitchThreeInputsDifferentRowGroupSize

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines
  from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    1. Subject is limited to 50 characters (not including Jira issue reference)
    1. Subject does not end with a period
    1. Subject uses the imperative mood ("add", not "adding")
    1. Body wraps at 72 characters
    1. Body explains "what" and "why", not "how"

### Style
- [ ] My contribution adheres to the code style guidelines and Spotless passes.
    - To apply the necessary changes, run `mvn spotless:apply -Pvector-plugins`

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain Javadoc that explain what it does

Commit:
add initial ParquetJoiner implementation

Commit:
add initial ParquetJoiner implementation

Commit:
Merge remote-tracking branch 'origin/master' into add-parquet-joiner

Commit:
refactor ParquetJoiner implementation

wgtmac commented:
cc @gszadovszky @ggershinsky @ConeyLiu in case you've missed the ML discussion: https://lists.apache.org/thread/5q7qhkl07of4ypc4t5zx3hyjgv12c7q2

gszadovszky commented:
I don't have the time to properly review this proposal but I like the idea.
I agree we shall move this joiner approach as close to the rewriter logic as possible. Maybe introducing new interfaces/implementations and use them where the logic is really different.

Commit:
extend the main test for multiple files on the right

Commit:
extend the main test for multiple files on the right

ConeyLiu commented:
The idea looks good to me too, actually, we have similar requirements to update history data. Feel free to ping me when this is ready for review.

Commit:
Merge branch 'master' into add-parquet-joiner

Commit:
converge join logic, crate a draft of options and rewriter

Commit:
move ParquetJoinTest logic to ParquetRewriterTest

Commit:
improve Parquet stitching test

Commit:
remove custom ParquetRewriter constructor

Commit:
remove custom ParquetRewriter constructor

Commit:
refactor ParquetRewriter

MaxNevermind commented:
@wgtmac 
Can I ask do a preliminary review of the code?
So far I've added a single test for a new feature. Wanted to add more tests and polish if the code looks good so far. All tests pass for both old and new code in ParquetRewriterTest atm. Will add more documentation to RewriteOptions if everything else looks good. Just fyi I've added an example of this new feature usage in a Gist here https://gist.github.com/MaxNevermind/0feaaf380520ca34c2637027ef349a7d.

wgtmac commented:
@MaxNevermind Sure, I will take a look this week.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -39,11 +36,20 @@
 
   private final ParquetConfiguration conf;
   private final List<InputFile> inputFiles;
+  private final List<List<InputFile>> inputFilesR;
```
Could we consolidate the internal member variables `inputFiles` and `inputFilesR`. My rough idea is to model these files into `HorizontalInputFile` and `VerticalInputFile`.
- VerticalInputFile: An `InputFile` of a single parquet file which may have partial or full schema.
- HorizontalInputFile: one or more `VerticalInputFile`s with same number rows but different schemas to form a logical wide file. It can choose one `VerticalInputFile` to be the left side and the others to be the right side in your join algorithm. If it contains only one `VerticalInputFile`, this is what we have today.

WDYT?

MaxNevermind replied:
So the RewriteOptions will contain a holder for List<VerticalInputFile> & List<HorizontalInputFile> or some sort of general List<StackableInputFile> that both VerticalInputFile & HorizontalInputFile need to implement? 


By the way. In my current implementation only the total number of rows of the left and on the right need to match. Individual files don't need the row count matched. It seem you suggest putting the same number of records into each VerticalInputFiles group. I'm not sure that particular requirement is easy to satisfy in practical environment. For example in Spark there are ways to preserve the ordering as I showed in the linked Gist in PR description but controlling the number of rows in each output file in Spark is not trivial as far as I know.

wgtmac replied:
That make sense. Files on the right side do not need to make number of rows to the left side.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +108,49 @@ public class ParquetRewriter implements Closeable {
+        .map(x -> getFileReaders(x, conf))
+        .collect(Collectors.toList());
+    ensureSameSchema(inputFiles);
+    inputFilesR.forEach(this::ensureSameSchema);
```
So this imposes strict schema check here. Is it possible to loose the restriction to allow the files on the right side to have different schemas but in total all columns are available? For example: a group of right side files with schema below:
- `a:bigint,b:string`, `c:double,d:binary`
- `a:bigint`, `b:string`, `c:double`, `d:binary`
- `a:bigint,b:string,c:double,d:binary`

I know this would be complex, just asking and not sure if it is an valid real world use case.

MaxNevermind replied:
In theory that is possible but that would make RewriteOptions for the right side more complex which is already not simple, it is List<List<InputFile>> right now, It would need to become List<List<List<InputFile>>> to allow something like that. Also while writing the right side the sate would need to become more complex too so it can track all of those files.

MaxNevermind replied:
On the second thought actually the right side might remain List<List<InputFile>>. The requirements are: the order of files for each column need to be expressed somehow to us and number of rows in for each column in total in all files must be the same as on the left side. Having those satisfied we can deduct which file has which column etc. Let me think a little more about it.

MaxNevermind replied:
I thought a little more about your suggestion. My notes:
- We can do that but we should not require the same number of rows for VerticalInputFiles in each HorizontalInputFile. In my opinion tools that produce Parquet files often use some threshold size in bytes(e.g.:HDFS block size) as a measure of next file spit. So it will be changeling for user to produce such files with matching row count.
- If we do that the state that tracks current file for each column on the right side will become more complex. Right now the state is easily maintained by a input file queue for each file group.
- I can see that we definitely will provide a more use-case coverage that way but I'm not sure if it will be useful. In my experience parquet files are usually produced using some widespread libraries/framework and those tools usually produce parquet dataset/tables which implies all output files have the same schema. I can see that somebody can produce such parquet files as you described programmatically, but in such cases those files unlikely to be large and the main use-case of my feature is in the context of optimization for very large datasets stitching, if there are no large datasets involved it might be easier to just completely rewrite files using full read & write.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +108,49 @@ public class ParquetRewriter implements Closeable {
 
-    // Init reader of the first input file
-    initNextReader();
+    extraMetaData.put(
```
I'm skeptical of the importance of metadata from the right side files. However I don't have a better suggestion here.

MaxNevermind replied:
Is there an action point here or we can resolve this?

wgtmac replied:
I'm a little concerned about the possible explosion of the metadata. Should we add an option to ignore those on the right side? Or at least add a comment to raise the potential issue.

MaxNevermind replied:
I thought about it and maybe we should make it as a general option for both left and right? In theory it might happen even without the right part as I understand. Let me know. If no, I will add an option for just the right part.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +108,49 @@ public class ParquetRewriter implements Closeable {
+    Stream.concat(inputFiles.stream(), inputFilesR.stream().flatMap(Collection::stream))
+        .forEach(x -> extraMetaData.putAll(x.getFileMetaData().getKeyValueMetaData()));
+
+    // TODO check that schema on the left and on the right is not identical
```
nit: the schema check now is complex so it worths a separate method.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -137,16 +160,31 @@ public ParquetRewriter(RewriteOptions options) throws IOException {
+      this.descriptorsMap =
+          schema.getColumns().stream().collect(Collectors.toMap(x -> ColumnPath.get(x.getPath()), x -> x));
+    } else {
+      this.descriptorsMap = schemaL.getColumns().stream()
```
OK, now I can answer my own question, the complex of loose schema enforcement on the right side may complicate `descriptorsMap` here.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -137,16 +160,31 @@ public ParquetRewriter(RewriteOptions options) throws IOException {
+        .map(x -> x.stream().mapToLong(ParquetFileReader::getRecordCount).sum())
+        .forEach(rowCountR -> {
+          if (rowCountL != rowCountR) {
+            throw new IllegalArgumentException("The number of records on the left and on the right don't match!");
```
nit: print the row count in the error msg.

MaxNevermind replied:
done

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -813,7 +841,7 @@ private Type extractField(GroupType candidate, Type targetField) {
         Type tempField = extractField(field.asGroupType(), targetField);
         if (tempField != null) {
-          return tempField;
+          return new GroupType(candidate.getRepetition(), candidate.getName(), tempField);
```
Is it an individual bug fix?

MaxNevermind replied:
yeah, that's the issue I mentioned in email communication, it become a problem when I started a using a different for the right side file group

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +932,208 @@ public byte[] getDictPageAAD() {
+      }
+
+      // Writing the right side
+      for (RightColumnWriter writer: columnWritersR) {
```
Is it possible to move this into `processBlocksFromReader`?

MaxNevermind replied:
Yeah, that would make sense, let me try.

MaxNevermind replied:
done

Commit:
apply spotless and address PR comments

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -110,6 +118,27 @@ public List<Path> getInputFiles() {
+   *
+   * @return a {@link List} of the associated right input {@link Path}s
+   */
+  public List<List<Path>> getInputFilesR() {
```
done

Commit:
move extra column writing into processBlocksFromReader

Commit:
add getInputFiles back

Commit:
Merge remote-tracking branch 'fork/master' into add-parquet-joiner

Commit:
fix extra ParquetRewriter constructor so tests can pass

Commit:
remove not needed TODOs

MaxNevermind commented:
@wgtmac 
Let me if you want me to implement your VerticalInputFile / HorizontalInputFile idea and what to do you see has left in this PR. 
Let me know what else I should do, I think I should add javadoc I can add more tests. I guess you are aware but just fyi ParquetRewriterTest already takes 3/4 of the total parquet-hadoop testing time, on my machine it is 16 min and 22 min correspondingly.

wgtmac commented:
Thanks for the update! I will take a look later this week.

> Let me if you want me to implement your VerticalInputFile / HorizontalInputFile idea and what to do you see has left in this PR.

No. The rows on both sides are different, so we don't have to bother with the concepts.

> I guess you are aware but just fyi ParquetRewriterTest already takes 3/4 of the total parquet-hadoop testing time, on my machine it is 16 min and 22 min correspondingly.

Yes, we probably should disable some parameter combinations.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +117,55 @@ public class ParquetRewriter implements Closeable {
+    });
+    List<Type> fields = Stream.concat(
+            fieldNamesL.values().stream()
+                .map(x -> fieldNamesR.getOrDefault(
```
Why do we need to try to get it from fieldNamesR?

MaxNevermind replied:
The idea behind is that I wanted to support the use-case when the left files have such a field but it has a different structure on the right.

wgtmac replied:
This implicit behavior may introduce unexpected use. Does it sound better to add a resolve strategy for duplicate column names? By default we can pick it up from the right side.

MaxNevermind replied:
> resolve strategy for duplicate

Can you please clarify what do you mean by that? You mean if the schema of the column on the left is "compatible" with a new column schema on the right side then we try to "merge" the columns?

wgtmac replied:
No, I mean we should be able to choose how to resolve columns with duplicate names, probable strategies are: pick left, pick right, fail.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -137,16 +175,34 @@ public ParquetRewriter(RewriteOptions options) throws IOException {
+          schema.getColumns().stream().collect(Collectors.toMap(x -> ColumnPath.get(x.getPath()), x -> x));
+    } else { // TODO: describe in documentation that only top level column can be overwritten
+      this.descriptorsMap = schemaL.getColumns().stream()
+          .filter(x -> x.getPath().length == 0 || !fieldNamesR.containsKey(x.getPath()[0]))
```
If we restrict both left and right sides do not have duplicate fields, would we make things simpler here and after?

MaxNevermind replied:
I wanted to preserve the original capabilities of ParquetRewriter which allowed duplicated columns as I understand, but for the right side that would create a problem and I added that check. Let me know if you want to support for duplicated columns in general.

wgtmac replied:
IIRC, the original rewriter does not support duplicated columns. However, supporting it while joining files is a good idea. Just as my previous comment, it would be better to support explicit resolve strategy for them if that is not too complicated.

ConeyLiu replied:
So, the left part columns' order could be changed if overwrite occurred.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+              .build();
+          CodecFactory codecFactory = new CodecFactory(new Configuration(), props.getPageSizeThreshold());
+          CompressionCodecFactory.BytesInputCompressor compressor =
+              codecFactory.getCompressor(chunk.getCodec());
```
This uses the same codec in the input file. Should we respect the trans-codec rewriter options?

wgtmac replied:
That makes sense. I didn't mean we have to implement everything.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -137,16 +175,34 @@ public ParquetRewriter(RewriteOptions options) throws IOException {
       }
 
       Set<ColumnPath> prunePaths = convertToColumnPaths(pruneColumns);
       schema = pruneColumnsInSchema(schema, prunePaths);
```
Right side columns get pruned here, but they are still written in the RightColumnWriter. Do you plan to enable pruning columns on the right side? For whatever reason, we need to state this in the docstring of RewriterOptions.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+        int dlvl = reader.getCurrentDefinitionLevel();
+        do {
+          if (dlvl < dMax) {
+            cWriter.writeNull(rlvl, dlvl);
```
If a column to mask (only nullifying column is supported) is on the right side, we also need to handle it here.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+              props.getAllocator(),
+              props.getColumnIndexTruncateLength(),
+              props.getPageWriteChecksumEnabled(),
+              writer.getEncryptor(),
```
IIUC, this line makes sure that right side column can be encrypted.

wgtmac reviewed (changes requested):
I do see that you've taken a lot of effort to consolidate the implementation with original rewriter features. Thank you for doing that! I've left some comments and I think we have made good progress.

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -119,6 +130,15 @@ public List<InputFile> getParquetInputFiles() {
+   *
+   * @return a {@link List} of the associated right {@link InputFile}s
+   */
+  public List<List<InputFile>> getParquetInputFilesR() {
```
Any preference?
- InputFilesRight
- InputFilesExtra
- InputFilesToJoin
- InputFilesStitched

wgtmac replied:
What about `InputFilesToJoinColumns` or `InputFilesToAddColumns`. This is not a strong opinion. Just want to clarify what to do with these input files.

ConeyLiu replied:
+1, `InputFilesR` should be replaced with a better name.

MaxNevermind replied:
renamed to InputFilesToJoinColumns

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+      cWriters.values().forEach(ColumnWriter::close);
+      for (ColumnDescriptor descriptor : descriptorsMap.values()) {
+        if (cPageStores.containsKey(descriptor))
+          cPageStores.get(descriptor).flushToFileWriter(writer);
```
> we may have some issues here since the column flush order does not match the schema order

I didn't know that this might cause problems. I though that schema column order and actual physical order are decoupled, I made my conclusion based on [Parquet File Metadata Schema](https://github.com/apache/parquet-format#metadata) as each column chunk stores its metadata internally: column path and file offset, so it seems there is no need for them to be coupled. Also that is weird that my new test didn't fail as I actually overwrite a column `FloatFraction` which is in the middle of the left schema, and as you can see from `processBlocksFromReader` method right side is always written at the end after all the left side is written. I need to investigate more.

wgtmac replied:
Yes, the specs do not require the column chunk order should be coupled to the schema order. However, I'm not sure if this will break some parquet implementations on the wild.

pyckle replied:
I encountered this same question while writing a parquet driver (https://github.com/Earnix/parquetforge). So I'll share what I dug up.
The parquet-format spec requires the ColumnChunk order in the footer to match the schema order. That's documented here: https://github.com/apache/parquet-format/blob/1dbc814b97c9307687a2e4bee55545ab6a2ef106/src/main/thrift/parquet.thrift#L1002
In the JIRA ticket linked to the commit message (now migrated to GH) - https://github.com/apache/parquet-java/issues/1734 it seems that at least impala accepts columns written in arbitrary orders, and ordering the Dict/DataPages of a column chunk is not required.
With this said, be aware that this can adversely affect performance. See: https://dl.acm.org/doi/10.1145/3035918.3035930

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+          rowGroupIdxOut++;
+          if (rowGroupIdxOut == blocks.size()) {
+            inputFiles.poll();
+            rowGroupIdxOut = 0;
```
You mean `rowGroupIdxOut++`? Yes, it is required. I can rewrite the code like this if it is easier to read:
```
        if (rowsToWrite > 0 || (block.getRowCount() == writtenFromBlock)) {
          if (rowGroupIdxR == blocks.size()) {
            inputFiles.poll();
            rowGroupIdxR = 0;
          } else {
            rowGroupIdxR++;
          }
          ...
        }
```

wgtmac replied:
I misunderstood the code, ignore it pls.

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+    }
+
+    public void writeRows(int rowGroupIdx, long rowsToWrite) throws IOException {
+      if (rowGroupIdxIn != rowGroupIdx) {
```
We can but right now that is not supposed to happen and I believe it would brake the right side tests immediately and a problem become visible through that. I can add something like that if you want me to:
```
      if (rowGroupIdxL > rowGroupIdx) {
        throw new IOException("A row group index decrease is determined in RightColumnWriter! Current index: " 
            + rowGroupIdxL + ", new index: " + rowGroupIdx);
      }
```
Let me know

wgtmac replied:
Yes, that sounds good.

MaxNevermind replied:
Should I add a proposed check in a message above or we can resolve this as is?

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +953,171 @@ public byte[] getDictPageAAD() {
+      initWriters();
+    }
+
+    public void writeRows(int rowGroupIdx, long rowsToWrite) throws IOException {
```
Done, renamed them to rowGroupIdxL & rowGroupIdxR, hope that improves readability 😄

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -347,8 +390,14 @@ private void processBlocksFromReader(IndexCache indexCache) throws IOException {
+
       writer.endBlock();
       numBlocksRewritten++;
+      rowGroupIdx++;
```
You are right. Replaced rowGroupIdx with numBlocksRewritten. We still need to pass it to RightColumnWriter because it is a static class.

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -347,8 +390,14 @@ private void processBlocksFromReader(IndexCache indexCache) throws IOException {
 
+      // Writing extra columns
+      for (RightColumnWriter writer : columnWritersR) {
+        writer.writeRows(rowGroupIdx, blockMetaData.getRowCount());
```
done

MaxNevermind commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -60,10 +63,7 @@
-import org.apache.parquet.hadoop.ColumnChunkPageWriteStore;
-import org.apache.parquet.hadoop.IndexCache;
-import org.apache.parquet.hadoop.ParquetFileWriter;
+import org.apache.parquet.hadoop.*;
```
done

Commit:
address PR comments

MaxNevermind commented:
@wgtmac 
Can you check out the comments I left please?

wgtmac reviewed (commented):
Sorry for the delay. Please see my new comments.

wgtmac commented:
@ConeyLiu Would you mind taking a look as well?

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +121,55 @@ public class ParquetRewriter implements Closeable {
+        .collect(Collectors.toList());
+    ensureSameSchema(inputFiles);
+    inputFilesR.forEach(this::ensureSameSchema);
     LOG.info("Start rewriting {} input file(s) {} to {}", inputFiles.size(), options.getParquetInputFiles(), out);
```
This log only contains left input files. It would be better to add the right files as well.

MaxNevermind replied:
done

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +955,175 @@ public byte[] getDictPageAAD() {
+    private final Map<ColumnDescriptor, ColumnWriter> cWriters = new HashMap<>();
+    private int rowGroupIdxL = 0; // index of the rowGroup of the current file on the left
+    private int rowGroupIdxR = 0; // index of the rowGroup of the current file on the right
+    private int writtenFromBlock = 0;
```
I suggest using long type.

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +955,175 @@ public byte[] getDictPageAAD() {
+        long leftInBlock = block.getRowCount() - writtenFromBlock;
+        long writeFromBlock = Math.min(rowsToWrite, leftInBlock);
+        for (ColumnChunkMetaData chunk : chunks) {
+          if (chunk.isEncrypted()) {
```
Why not check this at `initWriters`?

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +955,175 @@ public byte[] getDictPageAAD() {
+                  ? ParquetProperties.WriterVersion.PARQUET_2_0
+                  : ParquetProperties.WriterVersion.PARQUET_1_0;
+          ParquetProperties props = ParquetProperties.builder()
+              .withWriterVersion(writerVersion)
```
The output file could be mixed v1/v2 format if the part of the right files has mixed format. That's right?

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -904,4 +955,175 @@ public byte[] getDictPageAAD() {
+                  : ParquetProperties.WriterVersion.PARQUET_1_0;
+          ParquetProperties props = ParquetProperties.builder()
+              .withWriterVersion(writerVersion)
+              .withBloomFilterEnabled(bloomFilterLength > 0)
```
Does the bloom filter have the same mixed problem?

ConeyLiu commented on parquet-hadoop/src/test/java/org/apache/parquet/hadoop/rewrite/ParquetRewriterTest.java:
```diff
@@ -740,6 +858,30 @@ private MessageType createSchema() {
+  }
+
+  private MessageType createSchemaR2() {
+    return new MessageType("schema", new PrimitiveType(REPEATED, FLOAT, "FloatFraction"));
```
Could you add an overwrite column with a different type?

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -137,16 +179,34 @@ public ParquetRewriter(RewriteOptions options) throws IOException {
+        .map(x ->
+            x.stream().mapToLong(ParquetFileReader::getRecordCount).sum())
+        .forEach(rowCountR -> {
+          if (rowCountL != rowCountR) {
```
We only check the row count. How do we check the record order? That should be a key requirement as well.

ConeyLiu commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -115,19 +121,55 @@ public class ParquetRewriter implements Closeable {
+                    x.getName())) // takes fields on the right if it was not present on the left
+            )
+        .collect(Collectors.toList());
+    schema = new MessageType(schemaL.getName(), fields);
```
The schema does not match the actual column orders if there is any overwrite. We need to add ut to verify this.

MaxNevermind replied:
@wgtmac you also mentioned that same problem [in this comment](https://github.com/apache/parquet-mr/pull/1273/files#r1547172865).
I've just spend some time on trying to refactor this part. I tried to align schema column order and actual write order of columns and it is mostly doable but RightColumnWriter state becomes too complex, for example file queue need to be present for each individual column on the right. I think the code will become to difficult to support long term if we do that.

ConeyLiu commented:
Thanks @MaxNevermind for the great work and thanks @wgtmac for pinging me. I am sorry for the late reply.

Aleis007891 reviewed (changes requested):
Yy

Commit:
Merge remote-tracking branch 'origin/master' into add-parquet-joiner

Commit:
rename inputFilesR to inputFilesToJoin

Commit:
rename inputFilesR to inputFilesToJoinColumns

Commit:
add getParquetInputFiles listing to the rewrite start logging

MaxNevermind commented:
@wgtmac @ConeyLiu 

fyi
Wanted to share an idea: we can use binary copy on the right side, do not do read-write for right columns, that is possible if the number of rows and ordering in row groups on the left and on the right is the same. I believe a related ideas were discussed in this PR. I originally doubt possibility of user being able to produce files in such a fashion, but this week I’ve found a way to do that in Spark 3.3+. I can create a utility class([similar to the one from above](https://gist.github.com/MaxNevermind/0feaaf380520ca34c2637027ef349a7d)) that takes the input files, run transformations for them, and then write the results in a way that it preserves original input files’ names, row groups row count and ordering. The benefit are of course the speed because of binary copy and simplicity as we don't need RightColumnWriter. I will try to validate my idea the next week and get back with the result. If I’m able to do that I suggest to simplify this PR or maybe split the effort into two: one simpler version for simple binary copy and the second one based on current state of this PR.
